### ISSUE TITLE
unsound code patterns

### ISSUE THREAD

**Issue by Freax13** [coconut-svsm/svsm#104]
This codebase unfortunately uses code patterns that are unsound. A few examples (not exhaustive):
1. Improper synchronization:
https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/sev/ghcb.rs#L490-L508
`GHCBIOPort` should not implement `Sync` as it is not safe to share the contained `RefCell` across threads.
2. Unsafe use of raw pointers:
https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/console.rs#L13-L38
`Console::write_str` creates a reference to the point set in `Console::set` without making sure that the pointer is dereferencable.
3. Missing lifetimes:
https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/pagetable.rs#L710-L747
`PageTableRef` doesn't have a lifetime associated with it, so there's no guarantee that the contained `PageTable` pointer is still live when `PageTableRef::deref(_mut)` dereferences the pointer.

**Comment by joergroedel** [comment]
You are right, there are several C-isms still in the code from the early days of the SVSM, when I was still starting my Rust learning journey 😊

So for the whole console code the solution will be to simplify it once the VC handler is merged. Then we can remove all the logic that makes it need a link to the GHCB.

The page-table code needs a general cleanup, that will happen along with the virtual memory manager changes and conversion work. For now it is safe because all page-tables are static, but with modules/processes this will change.

If you see any other place that could be improved I am open for PRs or discussussions in this issue. There are several areas for cleanup I have in mind, too. One is, e.g., changing stage2 to not need the PerCpu implementation, which allows further simplifications in PerCpu code. But there are certainly more places.

**Comment by Freax13** [comment]
> If you see any other place that could be improved I am open for PRs or discussussions in this issue.

I might be able to take a shot at some of the issues over the weekend. 

> There are several areas for cleanup I have in mind, too. One is, e.g., changing stage2 to not need the PerCpu implementation, which allows further simplifications in PerCpu code. But there are certainly more places.

In that case it's probably a good idea to open issues to track these future tasks :)

**Comment by Freax13** [comment]
Continued list of issues:
- [`PageRef::new`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/alloc.rs#L701) should be unsafe. The [`AsMut<[u8; PAGE_SIZE]>`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/alloc.rs#L720-L725) impl is dubious given that `PageRef` implements [`Clone`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/alloc.rs#L727-L735)
- [`this_cpu_mut`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/cpu/percpu.rs#L498) is unsound as it allows creating multiple mutable references pointing to the same location.
- [`PerCpu::ghcb`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/cpu/percpu.rs#L346) shouldn't give out mutable references with a static lifetime.
- [`zero_mem_region`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/utils/util.rs#L53) should be unsafe.
- [`VmsaRef`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/cpu/percpu.rs#L87-L107) allows turning an arbitrary address into a mutable reference.
- Some of the methods of [`GuestPtr`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/guestmem.rs#L174-L223) should be unsafe. 
- [`write_u8`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/guestmem.rs#L44) should be unsafe.
- [`VMSA::from_virt_addr`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/sev/vmsa.rs#L292) should be unsafe.
- [`copy_secrets_page`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/sev/secrets_page.rs#L31) should be unsafe.
- [`ImmutAfterInitCell::reinit`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/utils/immut_after_init.rs#L145) should be unsafe (or removed altogether).
- [`ImmutAfterInitCell::check_init`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/utils/immut_after_init.rs#L94) should also check the init flag when debug assertions are disabled.
- [`load_idt`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/cpu/idt.rs#L139C25-L139C25) either needs to take a 'static reference or be unsafe. The same applies to [load_tss](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/cpu/gdt.rs#L32).

Other than these safety violations, there is a lot of unsafe code in general and a lot of it is not very contained. Some of that's unavoidable given the nature of low level kernel code, but my guess is that this can be improved. 
Ideally unsafe code should be used in self-contained abstractions that can then be used from safe code (i.e. `Vec<T>` hides all the gory unsafe details and exposes a mostly safe API). Right now there is a lot of code calling unsafe functions (or functions that should be unsafe) in other modules far away. This makes it more difficult to verify and test.

**Comment by 00xc** [comment]
> Continued list of issues:
> 
> * [`PageRef::new`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/alloc.rs#L701) should be unsafe. The [`AsMut<[u8; PAGE_SIZE]>`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/alloc.rs#L720-L725) impl is dubious given that `PageRef` implements [`Clone`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/alloc.rs#L727-L735)

Yeah this is definitely wrong. Opened #117 for this.

**Comment by 00xc** [comment]
> Continued list of issues:
> ...
> * Some of the methods of [`GuestPtr`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/guestmem.rs#L174-L223) should be unsafe.
> * [`write_u8`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/guestmem.rs#L44) should be unsafe.

These are fine in my opinion, as invalid accesses are properly recovered from via the exception table. I sort of mentioned this in the commit message for 0955d4e86 (`"mm/guestmem: remove unsafe from {read,write}_u8()"`).

**Comment by Freax13** [comment]
> > Continued list of issues:
> > ...
> > 
> > * Some of the methods of [`GuestPtr`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/guestmem.rs#L174-L223) should be unsafe.
> > * [`write_u8`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/mm/guestmem.rs#L44) should be unsafe.
> 
> These are fine in my opinion, as invalid accesses are properly recovered from via the exception table. I sort of mentioned this in the commit message for [0955d4e](https://github.com/coconut-svsm/svsm/commit/0955d4e86c457a16bfa9e88b3694856139a1944a) (`"mm/guestmem: remove unsafe from {read,write}_u8()"`).

Exceptions are not the only concern though, there are plenty of ways to cause havoc by writing to writable pages. For example these functions could be used to bypass the borrow checker or corrupt return pointers on the stack.

**Comment by joergroedel** [comment]
I suggest opening a new issue with the remaining points and keep track of the to-be-fixed places in a checklist in the first comment of the issue. This way we can track more easily what remains to be fixed.

**Comment by Freax13** [comment]
> I suggest opening a new issue with the remaining points and keep track of the to-be-fixed places in a checklist in the first comment of the issue. This way we can track more easily what remains to be fixed.

Sounds good to me. Chances are that the issues listed here are out of date anyway.

Currently, I don't have the time to do this but feel free to do so. If everything goes well, I'll have more time to contribute in about two months.

**Comment by p4zuu** [comment]
> I suggest opening a new issue with the remaining points and keep track of the to-be-fixed places in a checklist in the first comment of the issue. This way we can track more easily what remains to be fixed.

I have some time to spend atm, I'll take this and investigate the possible still-alive unsound issues. @Freax13 feel free to jump in when you have free time :)

**Comment by 00xc** [comment]
Some updates:

> * [`PerCpu::ghcb`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/cpu/percpu.rs#L346) shouldn't give out mutable references with a static lifetime.

I'm working on having borrow checking for the GHCB at the moment.

> * [`VmsaRef`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/cpu/percpu.rs#L87-L107) allows turning an arbitrary address into a mutable reference.

Will be fixed by #348.

> * [`VMSA::from_virt_addr`](https://github.com/coconut-svsm/svsm/blob/739fa7f423f9a19c0615e22d199e24493a0efe84/src/sev/vmsa.rs#L292) should be unsafe.

Outdated

**Comment by 00xc** [comment]
> I have some time to spend atm, I'll take this and investigate the possible still-alive unsound issues. @Freax13 feel free to jump in when you have free time :)

If you're up for it, could you assign yourself in the development document? You can add me as well, as I'm already working on a couple of the points above.

**Comment by p4zuu** [comment]
> > I have some time to spend atm, I'll take this and investigate the possible still-alive unsound issues. @Freax13 feel free to jump in when you have free time :)
> 
> If you're up for it, could you assign yourself in the development document? You can add me as well, as I'm already working on a couple of the points above.

I created #359 to track the remaining unsound patterns. The list still needs a bit of triage, we can discuss it in the issue. We could also maybe add a "Tracking" label to the new issue (or similar name) if we plan to append the new found patterns to the same list.
I also created #360 to set the owners to the three of us, and add the new tracking issue. I guess we can close this issue once the #360 is merged :)
